packages/fxa-auth-server/lib/routes/subscriptions.js

```diff
@@ -10,6 +10,597 @@ const ScopeSet = require('../../../fxa-shared').oauth.scopes;
+
+    // This call needs to succeed for us to consider this a success.
+    await this.db.createAccountSubscription({
+      uid,
+      subscriptionId,
+      productId: newProductId,
+      createdAt: Date.now(),
+    });
+
+    const devices = await request.app.devices;
+    await this.push.notifyProfileUpdated(uid, devices);
+    this.log.notifyAttachedServices('profileDataChanged', request, {
+      uid,
+      email,
+    });
+    await this.profile.deleteCache(uid);
+
+    return { subscriptionId };
+  }
+
+  async listPlans(request) {
+    this.log.begin('subscriptions.listPlans', request);
+    await handleAuth(this.db, request.auth);
+    const plans = await this.payments.allPlans();
+    return plans;
+  }
+
+  async listActive(request) {
+    this.log.begin('subscriptions.listActive', request);
+    const { uid } = await handleAuth(this.db, request.auth, true);
+    return this.db.fetchAccountSubscriptions(uid);
+  }
+
+  async getCustomer(request) {
+    this.log.begin('subscriptions.getCustomer', request);
+    const { uid, email } = await handleAuth(this.db, request.auth, true);
+    const customer = await this.payments.fetchCustomer(uid, email, [
+      'data.subscriptions.data.latest_invoice',
+    ]);
+    if (!customer) {
+      throw error.unknownCustomer(uid);
+    }
+    let response = { subscriptions: [] };
+    if (customer.sources && customer.sources.data.length > 0) {
+      // Currently assume a single source, and we can only access these attributes
+      // on cards.
+      const src = customer.sources.data[0];
+      if (src.object === 'card') {
+        response = {
+          ...response,
+          payment_type: src.funding,
+          last4: src.last4,
+          exp_month: src.exp_month,
+          exp_year: src.exp_year,
+        };
+      }
+    }
+
+    response.subscriptions = await this.payments.subscriptionsToResponse(
+      customer.subscriptions
+    );
+    return response;
+  }
+}
+
+const directRoutes = (log, db, config, customs, push, mailer, profile) => {
+  const payments = new stripe(log, config);
+  const directStripeRoutes = new DirectStripeRoutes(
+    log,
+    db,
+    config,
+    customs,
+    push,
+    mailer,
+    profile,
+    payments
+  );
+
+  // FIXME: All of these need to be wrapped in Stripe error handling
+  // FIXME: Many of these stripe calls need retries with careful thought about
+  //        overall request deadline. Stripe retries must include a idempotency_key.
+  return [
+    {
+      method: 'GET',
+      path: '/oauth/subscriptions/clients',
+      options: {
+        auth: {
+          payload: false,
+          strategy: 'subscriptionsSecret',
+        },
+        response: {
+          schema: isA.array().items(
+            isA.object().keys({
+              clientId: isA.string(),
+              capabilities: isA.array().items(isA.string()),
+            })
+          ),
+        },
+      },
+      handler: request => directStripeRoutes.getClients(request),
```

Comment:
Curious what folks think of this pattern for getting handlers out of this hapi data structure - and if we might want to eventually extend it into reorgs of other routes?